### DISCUSSION TITLE
Question about event handling with IPRoute API

### DISCUSSION THREAD

**Discussion by ghost** [svinota/pyroute2#877] [Q&A]
Hello,

first of all I apologize for using the bugtracker as a means to get information. Didn't want to mail Peter directly, because maybe someone else might have a good hint for me.

Anyway, I'm trying to use v0.5.18 of pyroute2 to set routing rules (in the policy based routing sense) when the configuration of an interface changes. By change I mean adding of an IP address, update, remove, etc.

I'm pretty sure that the IPRoute API/class provides everything that is necessary to set the routing rules, but I'm unsure about how to implement the monitoring loop. I checked the examples, and `ip_monitor.py` looks like the kind of thing I want. OK, there's no looping here, but I guess that would be trivial. However the example uses a IPRSocket object, and as far as I understand, if I'm using that one, that I have to do pretty much everything myself.

So basically what I'm looking for is something like:
```
with IPRoute() as ipr:
    while ipr.wait_for_event():
        evt = ipr.get_event()
        manage_my_routing_rules(evt)
```

So ipr actually has a `get()` method, but apparantly it doesn't do what I want. Also there's `register_callback()`, which could be potentially be interesting. But I don't get my registered callback function to fire.

I guess I'm pretty much lost here. I'm grateful for any kind of pointers. Maybe there's an example of a one of the tests that is closer to what I want to achieve.

Thanks in advance!

With best wishes,
Tobias

**Comment by crosser** [comment]
I get `nlsock = IPRoute()`, obtain its file descriptor `nlfd = nlsock.fileno()`, add it to the poll set, and when `poll()` notifies about a read event on the descriptor, run `nlsock.get()` (in a loop) to read all messages.

**Comment by ghost** [comment]
Hello Eugene,

but isn't that pretty much functionally equivalent to just doing this?

```
with IPRoute() as ipr:
    while True:
        ipr.get()
```

I'm asking because this doesn't work for me. While running this, `ip monitor` shows some netlink events occuring. But I don't get any output from the script.

**Comment by ghost** [comment]
OK, so maybe I should've read the documentation more carefully.

This works for me:
```
with pyroute2.IPRoute() as ipr:
    ipr.bind()
    while True:
        for message in ipr.get():
            print(message)
```

**EDIT**:
So just to make sure. If I do things like that, I can just use IPRoute normally? So I'm not somehow interferring with netlink messages being processed by the library?

**Comment by svinota** [comment]
@tobiasjakobi-compleo 

> So I'm not somehow interferring with netlink messages being processed by the library?

Long story short: no, you will not interfere

tl;dr

1. every netlink request is marked with a unique `sequence_number` in the `header`; the responses come with the same number, so you never will confuse responses from different requests, even if they come overlapping in time

2. the library takes care of that, so you don't need to analyze `sequence_number` manually

3. all the broadcast messages come with `sequence_number` 0, and `ipr.get()` is actually `ipr.get(msg_seq=0)`

PS: converted the issue to a discussion

**Reply by ghost** [reply]
Hello Peter,

thanks for explanation. So, even if I retrieve the broadcast messages from the object, the library still evaluates them? In the sense, that updates reflected in the broadcast messages are applied to the object's internal state?

Or maybe the library doesn't even need the broadcast messages to keep its internal state in sync? Sorry, if I'm asking something that's probably obvious.

-- Tobias

**Reply by svinota** [reply]
The `IPRoute` object has no internal state except of tracing the responses, thus nothing to keep in sync. Essentially `IPRoute` is nothing but a simple socket object with ordinary `recv()` and `sendto()` methods. And with an FD that you can use in poll/select. The only thing the `IPRoute` implements is a higher level API that allows not to compile binary data manually.

Like you want to list all the interfaces in the system. You can manually build a byte array according to the RTNL protocol, open a netlink socket and send this data, receive the answer and parse it. Or you can use `ipr.link("dump")` that does all this for you.

So you can safely run `ipr.bind()` and any subsequent `ipr.get()` will return the next broadast — and it will not interfere with other calls like `ipr.link("dump")` in, for example, another thread. Moreover, `ipr.put()` and `ipr.get()` are thread-safe, so there will be no corrupted state even if you call the multiple `ipr.link("dump")` simultaneously on the same object.

More details — https://docs.pyroute2.org/iproute.html

…

There is another component in the library, even higher level API `NDB`, and it has the internal state, but it's a completely different story — https://docs.pyroute2.org/ndb.html

**Reply by ghost** [reply]
Thanks, I see now where my confusion stems from. I was under the impression that `IPRoute` was also somehow caching state, and that doing your aforementioned `ipr.link("dump")` would first look for the necessary information in the cache. Looks like I was mixing up here information about `IPRoute`, `NBD` and also libnl (which also does the caching thing).

**Reply by svinota** [reply]
No, there is no implicit state caching in `IPRoute`. It may be bad from the performance perspective, especially in our case, using Python, but it is a simple socket object with almost no tricks. The only trick we have in `IPRoute` is the userspace cache of received but not yet parsed messages, running in a separate thread, so there is no socket buffer overflow error — but it's also off by default.

In some cases using `NBD` may give more performance, since it's implementing the internal state, that's being updated in the background from broadcast messages. But it depends on the architecture and requirements.
